src/main/java/org/elasticsearch/index/gateway/IndexShardGateway.java

```diff
@@ -151,41 +151,49 @@ public void recover(boolean indexShouldExists, RecoveryState recoveryState) thro
             typesToUpdate = indexShard.performTranslogRecovery();
 
             indexShard.finalizeRecovery();
+            for (Map.Entry<String, Mapping> entry : typesToUpdate.entrySet()) {
+                validateMappingUpdate(entry.getKey(), entry.getValue());
+            }
             indexShard.postRecovery("post recovery from gateway");
         } catch (EngineException e) {
             throw new IndexShardGatewayRecoveryException(shardId, "failed to recovery from gateway", e);
         } finally {
             indexShard.store().decRef();
         }
-        for (final String type : typesToUpdate) {
-            final CountDownLatch latch = new CountDownLatch(1);
-            mappingUpdatedAction.updateMappingOnMaster(indexService.index().name(), indexService.mapperService().documentMapper(type), indexService.indexUUID(), new MappingUpdatedAction.MappingUpdateListener() {
-                @Override
-                public void onMappingUpdate() {
-                    latch.countDown();
-                }
+    }
 
-                @Override
-                public void onFailure(Throwable t) {
-                    latch.countDown();
-                    logger.debug("failed to send mapping update post recovery to master for [{}]", t, type);
-                }
-            });
-            cancellableThreads.execute(new CancellableThreads.Interruptable() {
-                @Override
-                public void run() throws InterruptedException {
-                    try {
-                        if (latch.await(waitForMappingUpdatePostRecovery.millis(), TimeUnit.MILLISECONDS) == false) {
-                            logger.debug("waited for mapping update on master for [{}], yet timed out", type);
+    private void validateMappingUpdate(final String type, Mapping update) {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final Throwable[] error = new Throwable[1];
+        mappingUpdatedAction.updateMappingOnMaster(indexService.index().name(), indexService.indexUUID(), type, update, new MappingUpdatedAction.MappingUpdateListener() {
+            @Override
+            public void onMappingUpdate() {
+                latch.countDown();
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+                latch.countDown();
+                error[0] = t;
+            }
+        });
+        cancellableThreads.execute(new CancellableThreads.Interruptable() {
+            @Override
+            public void run() throws InterruptedException {
+                try {
+                    if (latch.await(waitForMappingUpdatePostRecovery.millis(), TimeUnit.MILLISECONDS) == false) {
```

Comment:
waitForMappingUpdatePostRecovery defaults to 30s, give that the timeout now results in a failed shard (good!) I think we should be more lenient. Especially given the fact that local gateway recovery runs on full cluster restart where the master might be overloaded by things to do. How about something very conservative like 15m (which is what we use for the same update mapping  - see RecoverySourceHandler#updateMappingOnMaster)